packages/gatsby-core-utils/src/service-lock.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Service lock: handles service discovery for Gatsby develop processes
+ * The problem:  the develop process starts a proxy server, the actual develop process and a websocket server for communication. The two latter ones have random ports that need to be discovered. We also cannot run multiple of the same site at the same time.
+ * The solution: lockfiles! We create a lockfolder in `.config/gatsby/sites/${sitePathHash} and then write a file to that lockfolder for every service with its port.
+ */
+import path from "path"
+import os from "os"
+import crypto from "crypto"
+import lockfile from "proper-lockfile"
+import fse from "fs-extra"
+
+const globalConfigPath =
+  process.env.XDG_CONFIG_HOME || path.join(os.homedir(), `.config`)
+
+const hashString = (str: string): string =>
+  crypto.createHash(`md5`).update(str).digest(`hex`)
+
+const getLockfileDir = (programPath: string): string => {
+  const hash = hashString(programPath)
+
+  return path.join(globalConfigPath, `gatsby`, `sites`, `${hash}.lock`)
+}
+
+export const createServiceLock = async (
+  programPath: string,
+  name: string,
+  content: string
+): Promise<void> => {
+  const lockfileDir = getLockfileDir(programPath)
+
+  await fse.ensureDir(lockfileDir)
+
+  try {
+    await lockfile.lock(lockfileDir)
+  } catch (err) {
+    // TODO: Nice helpful error message
+    throw new Error(`The develop process is already running on another port.`)
```

Comment:
You could use the `report` function that is already used inside develop-process.ts

Reply:
Ah I actually need to rework this entirely, great catch!